File: android/app/src/main/kotlin/com/musicmuni/vozos/demo/sonix/PlaybackSection.kt
package com.musicmuni.vozos.demo.sonix

import android.content.Context
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.musicmuni.vozos.sonix.SonixPlayer
import com.musicmuni.vozos.demo.components.OptionChip
import io.github.aakira.napier.Napier
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream

@Composable
fun PlaybackSection(context: Context) {
    val scope = rememberCoroutineScope()

    var player by remember { mutableStateOf<SonixPlayer?>(null) }
    var isPlaying by remember { mutableStateOf(false) }
    var currentTimeMs by remember { mutableLongStateOf(0L) }
    var durationMs by remember { mutableLongStateOf(0L) }
    var volume by remember { mutableFloatStateOf(0.8f) }
    var pitch by remember { mutableFloatStateOf(0f) }
    var loopCount by remember { mutableIntStateOf(1) }
    var status by remember { mutableStateOf("Ready") }
    var isLoaded by remember { mutableStateOf(false) }

    // Initialize player
    LaunchedEffect(Unit) {
        try {
            status = "Loading..."

            // Copy asset to file on IO dispatcher
            val assetFile = withContext(Dispatchers.IO) {
                copyAssetToFile(context, "sample.m4a")
            }

            // Create player using Builder pattern with callbacks
            val newPlayer = SonixPlayer.Builder()
                .source(assetFile.absolutePath)
                .volume(volume)
                .pitch(pitch)
                .loopCount(loopCount)
                .onComplete {
                    Napier.d("Playback completed!")
                }
                .onLoopComplete { loopIndex, totalLoops ->
                    Napier.d("Loop $loopIndex of $totalLoops completed")
                }
                .onError { error ->
                    Napier.e("Playback error: $error")
                    status = "Error: $error"
                }
                .build()

            player = newPlayer
            durationMs = newPlayer.duration
            isLoaded = true
            status = "Loaded: sample.m4a"
            Napier.d("Audio loaded, duration: $durationMs ms")

            // Observe playback state
            launch {
                newPlayer.isPlaying.collect { playing ->
                    isPlaying = playing
                }
            }

            launch {
                newPlayer.currentTime.collect { time ->
                    currentTimeMs = time
                }
            }
        } catch (e: Exception) {
            Napier.e("Player init failed", e)
            status = "Error: ${e.message}"
        }
    }

    // Cleanup
    DisposableEffect(Unit) {
        onDispose {
            player?.release()
        }
    }

    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text(
            text = "Playback (SonixPlayer.Builder)",
            style = MaterialTheme.typography.titleMedium
        )

        Text(text = "Status: $status", style = MaterialTheme.typography.bodySmall)

        // Time display
        Text(
            text = "${formatTime(currentTimeMs)} / ${formatTime(durationMs)}",
            style = MaterialTheme.typography.bodyMedium
        )

        // Seek slider
        Slider(
            value = if (durationMs > 0) currentTimeMs.toFloat() / durationMs else 0f,
            onValueChange = { fraction ->
                val seekPos = (fraction * durationMs).toLong()
                player?.seek(seekPos)
            },
            enabled = isLoaded,
            modifier = Modifier.fillMaxWidth()
        )

        // Playback controls
        Row(
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Button(
                onClick = { player?.play() },
                enabled = isLoaded && !isPlaying
            ) {
                Text("Play")
            }
            Button(
                onClick = { player?.pause() },
                enabled = isPlaying
            ) {
                Text("Pause")
            }
            Button(
                onClick = {
                    player?.stop()
                    currentTimeMs = 0
                },
                enabled = isLoaded
            ) {
                Text("Stop")
            }
        }

        // Volume control
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text("Volume:", modifier = Modifier.width(60.dp))
            Slider(
                value = volume,
                onValueChange = {
                    volume = it
                    player?.let { p -> p.volume = it }
                },
                modifier = Modifier.weight(1f)
            )
            Text("${(volume * 100).toInt()}%", modifier = Modifier.width(40.dp))
        }

        // Pitch control - realtime updates
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text("Pitch:", modifier = Modifier.width(60.dp))
            Slider(
                value = (pitch + 12f) / 24f,
                onValueChange = {
                    pitch = (it * 24f) - 12f
                    player?.let { p -> p.pitch = pitch }
                },
                modifier = Modifier.weight(1f)
            )
            Text("${pitch.toInt()} st", modifier = Modifier.width(40.dp))
        }

        // Loop count
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text("Loop:", modifier = Modifier.width(60.dp))
            listOf(1, 2, 3, -1).forEach { count ->
                OptionChip(
                    selected = loopCount == count,
                    onClick = {
                        loopCount = count
                        player?.let { p -> p.loopCount = count }
                    },
                    label = if (count == -1) "∞" else count.toString()
                )
            }
        }

        // Fade controls
        Row(
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Button(
                onClick = {
                    scope.launch {
                        player?.fadeIn(1.0f, 1000)
                    }
                },
                enabled = isLoaded
            ) {
                Text("Fade In")
            }
            Button(
                onClick = {
                    scope.launch {
                        player?.fadeOut(1000)
                    }
                },
                enabled = isLoaded
            ) {
                Text("Fade Out")
            }
        }
    }
}

private fun formatTime(ms: Long): String {
    val seconds = (ms / 1000) % 60
    val minutes = (ms / 1000) / 60
    return String.format("%d:%02d", minutes, seconds)
}

fun copyAssetToFile(context: Context, assetName: String): File {
    val file = File(context.cacheDir, assetName)
    if (!file.exists()) {
        context.assets.open(assetName).use { input ->
            FileOutputStream(file).use { output ->
                input.copyTo(output)
            }
        }
    }
    return file
}

File: demo-apps/ios/VozOSDemo/VozOSDemo/Sections/Sonix/PlaybackSection.swift
import SwiftUI
import VozOS

struct PlaybackSection: View {
    @State private var player: SonixPlayer?
    @State private var isPlaying = false
    @State private var currentTimeMs: Int64 = 0
    @State private var durationMs: Int64 = 0
    @State private var volume: Float = 0.8
    @State private var pitch: Float = 0
    @State private var loopCount: Int32 = 1
    @State private var status = "Ready"
    @State private var isLoaded = false
    @State private var playingTask: Task<Void, Never>?
    @State private var timeTask: Task<Void, Never>?

    // Break out bindings to help Swift type checker
    private var seekBinding: Binding<Double> {
        Binding(
            get: { durationMs > 0 ? Double(currentTimeMs) / Double(durationMs) : 0 },
            set: { newValue in
                let seekPos = Int64(newValue * Double(durationMs))
                player?.seek(positionMs: seekPos)
            }
        )
    }

    private var volumeBinding: Binding<Double> {
        Binding(
            get: { Double(volume) },
            set: { newValue in
                volume = Float(newValue)
                player?.volume = volume
            }
        )
    }

    private var pitchBinding: Binding<Double> {
        Binding(
            get: { Double(pitch + 12) / 24 },
            set: { newValue in
                // Only update the local state - pitch is applied in onEditingChanged
                pitch = Float(newValue * 24 - 12)
            }
        )
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Playback (SonixPlayer.Builder)")
                .font(.headline)

            Text("Status: \(status)")
                .font(.caption)
                .foregroundColor(.secondary)

            // Time display
            Text("\(formatTime(currentTimeMs)) / \(formatTime(durationMs))")
                .font(.body)

            // Seek slider
            Slider(value: seekBinding)
                .disabled(!isLoaded)

            // Playback controls
            playbackControls

            // Volume control
            volumeControl

            // Pitch control
            pitchControl

            // Loop count
            loopControl

            // Fade controls
            fadeControls
        }
        .task {
            await initializePlayer()
        }
        .onDisappear {
            cancelObservers()
            player?.release()
        }
    }

    private var playbackControls: some View {
        HStack(spacing: 12) {
            Button("Play") {
                player?.play()
            }
            .disabled(!isLoaded || isPlaying)
            .buttonStyle(.borderedProminent)

            Button("Pause") {
                player?.pause()
            }
            .disabled(!isPlaying)
            .buttonStyle(.bordered)

            Button("Stop") {
                player?.stop()
                currentTimeMs = 0
            }
            .disabled(!isLoaded)
            .buttonStyle(.bordered)
        }
    }

    private var volumeControl: some View {
        HStack {
            Text("Volume:")
                .frame(width: 60, alignment: .leading)
            Slider(value: volumeBinding)
            Text("\(Int(volume * 100))%")
                .frame(width: 40)
        }
    }

    private var pitchControl: some View {
        HStack {
            Text("Pitch:")
                .frame(width: 60, alignment: .leading)
            Slider(
                value: pitchBinding,
                onEditingChanged: { editing in
                    if !editing {
                        // Only apply pitch when drag ends for responsiveness
                        player?.pitch = pitch
                    }
                }
            )
            Text("\(Int(pitch)) st")
                .frame(width: 40)
        }
    }

    private var loopControl: some View {
        HStack {
            Text("Loop:")
                .frame(width: 60, alignment: .leading)
            loopButton(count: 1)
            loopButton(count: 2)
            loopButton(count: 3)
            loopButton(count: -1)
        }
    }

    @ViewBuilder
    private func loopButton(count: Int) -> some View {
        let label = count == -1 ? "inf" : "\(count)"
        if count == Int(loopCount) {
            Button(label) {
                loopCount = Int32(count)
                player?.loopCount = loopCount
            }
            .buttonStyle(.borderedProminent)
            .controlSize(.small)
        } else {
            Button(label) {
                loopCount = Int32(count)
                player?.loopCount = loopCount
            }
            .buttonStyle(.bordered)
            .controlSize(.small)
        }
    }

    private var fadeControls: some View {
        HStack(spacing: 12) {
            Button("Fade In") {
                Task {
                    try? await player?.fadeIn(targetVolume: 1.0, durationMs: 1000)
                }
            }
            .disabled(!isLoaded)
            .buttonStyle(.bordered)
            .controlSize(.small)

            Button("Fade Out") {
                Task {
                    try? await player?.fadeOut(durationMs: 1000)
                }
            }
            .disabled(!isLoaded)
            .buttonStyle(.bordered)
            .controlSize(.small)
        }
    }

    private func initializePlayer() async {
        do {
            status = "Loading..."

            // Copy asset to file for loading
            guard let assetPath = copyAssetToFile(name: "sample", ext: "m4a") else {
                status = "Asset not found"
                return
            }

            // Create player using Builder pattern with callbacks
            let newPlayer = try await SonixPlayer.Builder()
                .source(path: assetPath)
                .volume(volume: volume)
                .pitch(semitones: pitch)
                .loopCount(count: Int32(loopCount))
                .onComplete {
                    print("Playback completed!")
                }
                .onLoopComplete { loopIndex, totalLoops in
                    print("Loop \(loopIndex) of \(totalLoops) completed")
                }
                .onError { error in
                    print("Playback error: \(error)")
                }
                .build()

            player = newPlayer
            durationMs = newPlayer.duration
            isLoaded = true
            status = "Loaded: sample.m4a"

            // Observe playback state
            observePlaybackState(newPlayer)
        } catch {
            status = "Error: \(error.localizedDescription)"
        }
    }

    private func observePlaybackState(_ player: SonixPlayer) {
        // Use type-safe observers instead of force casts
        playingTask = player.observeIsPlaying { self.isPlaying = $0 }
        timeTask = player.observeCurrentTime { self.currentTimeMs = $0 }
    }

    private func cancelObservers() {
        playingTask?.cancel()
        timeTask?.cancel()
    }

    private func formatTime(_ ms: Int64) -> String {
        let seconds = (ms / 1000) % 60
        let minutes = (ms / 1000) / 60
        return String(format: "%d:%02d", minutes, seconds)
    }
}

/// Helper to copy asset to temp file
func copyAssetToFile(name: String, ext: String) -> String? {
    guard let assetURL = Bundle.main.url(forResource: name, withExtension: ext) else {
        return nil
    }
    let tempDir = FileManager.default.temporaryDirectory
    let destURL = tempDir.appendingPathComponent("\(name).\(ext)")

    if !FileManager.default.fileExists(atPath: destURL.path) {
        do {
            try FileManager.default.copyItem(at: assetURL, to: destURL)
        } catch {
            print("Failed to copy asset: \(error)")
            return nil
        }
    }
    return destURL.path
}

#Preview {
    PlaybackSection()
        .padding()
}
